tests/unit/aofrw.tcl

```diff
@@ -17,10 +17,12 @@ start_server {tags {"aofrw external:skip logreqres:skip"} overrides {save {}}} {
             set master_host [srv 0 host]
             set master_port [srv 0 port]
             set load_handle0 [start_write_load $master_host $master_port 10]
-            set load_handle1 [start_write_load $master_host $master_port 10]
-            set load_handle2 [start_write_load $master_host $master_port 10]
-            set load_handle3 [start_write_load $master_host $master_port 10]
-            set load_handle4 [start_write_load $master_host $master_port 10]
+            # FIXME: Temporary disabling some load to prevent this test case
+            # from hanging indefinitely. This needs to be investivaged.
+            #set load_handle1 [start_write_load $master_host $master_port 10]
+            #set load_handle2 [start_write_load $master_host $master_port 10]
+            #set load_handle3 [start_write_load $master_host $master_port 10]
+            #set load_handle4 [start_write_load $master_host $master_port 10]
```

Comment:
Without commenting out these lines, this test case is hundreds of times slower than on unstable.

The test case sends random SET commands using multiple write load workers, creating millions of keys and performs AOF rewrite in the background. Then it reloads the AOF using `DEBUG LOADAOF`. It's the loading of AOF without preamble that takes time, i.e. loading by calling the `setCommand()` function. Each SET command takes ~300 microseconds (100 for lookupKey and 200 for setKey), compared to ~1 microsecond before the DEBUG RELOAD and on unstable.

When debugging, I discovered very long probing chains, like length 6000 or so here, possibly spanning all non-empty buckets in the table.

AOF rewrite creates the AOF by iterating the keys hashset. The iterator walks in cursor/probing order. When these keys are loaded in that same order, the first keys all hash to the same bucket and create long probing chains. I tried changing the iterator to just increment normally (index++). It gets slightly better but I still see long probing chains. (Not exactly sure why yet.)

The reason we don't see this slowdown in the preamble case is RDB contains a size and we expand the hash table in advance when loading an RDB.

Anyway, I believe @madolson's idea from https://github.com/valkey-io/valkey/pull/1186#pullrequestreview-2378600726 can solve this:

> instead of using the `everfull` bit to start probing, we could use it to indicate the last position is a pointer to another bucket and do linked list resolution. This does a couple of things:
>
> 1. Prevents significant chain lengths, since collisions won't cause issues.

Reply:
Solved by using bucket-chaining instead of open addressing.